src/test/kotlin/com/shortener/service/UrlServiceTest.kt

```diff
@@ -0,0 +1,175 @@
+        assertThat(response.shortenedUrl).isNotNull
+        assertThat(response.shortenedUrl).isNotEqualTo(goodUrl)
+    }
+
+    /**
+     * Verifies that two requests for shortening the same url does not produce identical shortened urls
+     */
+    @Test
+    fun shortenSameUrlSuccessTest() {
+        val request1 = UrlShortenRequest(goodUrl)
+        val response1 = urlService.shortenUrl(request1)
+
+        val request2 = UrlShortenRequest(goodUrl)
+        val response2 = urlService.shortenUrl(request2)
+
+        assertThat(response1.shortenedUrl).isNotEqualTo(response2.shortenedUrl)
+    }
+
+    /**
+     * Verifies that an input url with no schema can still be converted
+     */
+    @Test
+    fun shortenUrlWithNoSchemaSuccessTest() {
+        val request = UrlShortenRequest(goodUrlNoSchema)
+        val response = urlService.shortenUrl(request)
+        assertThat(response).isNotNull
+        assertThat(response.shortenedUrl).isNotNull
+        assertThat(response.shortenedUrl).isNotEqualTo(goodUrlNoSchema)
+    }
+
+    /**
+     * Tests that an invalid url cannot be converted, and exception is thrown
+     */
+    @Test
+    fun shortenUrlBadUrlFailedTest() {
+        val request = UrlShortenRequest(badUrl)
+        val exception = assertThrows<InvalidInputUrl> { urlService.shortenUrl(request) }
+
+        assertInvalidInputUrlMessage(exception)
+    }
+
+    /**
+     * Tests that an invalid url, lacking schema, cannot be converted, and exception is thrown
+     */
+    @Test
+    fun shortenUrlBadUrlNoSchemaFailedTest() {
+        val request = UrlShortenRequest(badUrlNoSchema)
+        val exception = assertThrows<InvalidInputUrl> { urlService.shortenUrl(request) }
+        assertInvalidInputUrlMessage(exception)
+    }
+
+    /**
+     * Tests that a shortened url can be reversed to the initial input url
+     */
+    @Test
+    fun decodeUrlSuccessTest() {
+        val encodedSequence = "a1b2"
+        var entry = createUrlEntry(goodUrl, false)
+        entry = urlEntryRepository.save(entry)
+        val decodedEntryId = entry.getId()
+
+        Mockito.`when`(idEncoder.decode(encodedSequence)).thenReturn(longArrayOf(decodedEntryId!!))
+
+        val response = urlService.decodeUrl(encodedSequence)
+        assertThat(response).isNotNull
+        assertThat(response).isEqualTo(goodUrl)
+    }
+
+    /**
+     * Verifies that a nonexistent shortened url cannot be converted, and exception is thrown
+     */
+    @Test
+    fun decodeUrlInvalidSequenceTest() {
+        val encodedSequence = "a1b2"
+        Mockito.`when`(idEncoder.decode(encodedSequence)).thenReturn(longArrayOf())
+
+        val exception = assertThrows<InvalidInputUrl> { urlService.decodeUrl(encodedSequence) }
+        assertInvalidShortenedInputUrlMessage(exception)
+    }
+
+    /**
+     * Verifies that a shortened url which has expired cannot be converted, and exception is thrown
+     */
+    @Test
+    fun decodeUrlExpiredUrlTest() {
```

Comment:
fun `Verifies that a shortened url which has expired cannot be converted, and exception is thrown`() {} and you can delete the comments

Reply:
Done